src/p11_wrap.cpp

```diff
@@ -0,0 +1,274 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2022 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+/*
+ * This file is the only place where we should see any
+ * methods from LibP11. Any other code should use
+ * the more high-level methods declared and define in
+ * p11_wrap.cpp and p11_wrap.h.
+ *
+ */
+
+#include <boost/algorithm/hex.hpp>
+#include <cassert>
+#include <cstddef> /* this has to come before cppc (bug in boost) */
+#include <exception>
+
+#include <cppc/checkcall.hpp>
+#include "mococrw/openssl_wrap.h"
+#include "mococrw/p11_wrap.h"
+
+namespace mococrw
+{
+namespace p11
+{
+std::string P11Exception::generateP11ErrorString()
+{
+    // Note, pkcs11_CTX_new() internally loads LibP11 error strings,
+    // i.e., it invokes ERR_load_PKCS11_strings(). We therefore
+    // don't trigger the loading ourselves.
+    auto error = openssl::lib::OpenSSLLib::SSL_ERR_get_error();
```

Comment:
Do we really get p11 errors via OpenSSL? Why is this here then a duplication of the existing OpenSSL error class?

Reply:
Indeed! LibP11 registers error strings via OpenSSL's ERR_load_strings() function. See https://github.com/OpenSC/libp11/blob/d5e07497e915c339f4c9e7534750eb4d8309b99b/src/p11_err.c#L68